plum/parametric.py

```diff
@@ -30,8 +30,37 @@
 
 _dispatch = Dispatcher()
 
+class ParametricTypeMeta(TypeMeta):
+    """Parametric Types must be instantiated with indexing."""
 
-class CovariantMeta(TypeMeta):
+    def __getitem__(self, *p):
+        # Type[Par1, Par2] creates a new parametric type
+
+        return self.__new__(self, *p)
+
+    def __call__(cls, *args, **kwargs):
+        # Type(arg1, arg2, kwargs) will first construct the
+        # parametric subtype T = Type[type(arg1), type(arg2)]
+        # and then call the equivalent of T(arg1, arg2, **kwargs)
+
+        if not cls.is_concrete:
+            argsT = tuple(type(arg) for arg in args)
```

Comment:
If `len(args) == 1`, is `argsT = args[0]` a sensible simplification? Moreover, shall we name `argsT` to `type_parameters` to slightly clarify?

Reply:
I agree with the renaming.
However I don't understand the semplification you are proposing. This code works for an arbitrary number of parameters

Reply:
Currently, this is what happens:

```python
In [1]: from plum import parametric

In [2]: @parametric
   ...: class A:
   ...:     def __init__(self, *xs):
   ...:         pass
   ...:

In [3]: A(1)
Out[3]: <__main__.A[(<class 'int'>,)] at 0x7fbfa809bf60>

In [4]: A[int](1)
Out[4]: <__main__.A[<class 'int'>] at 0x7fbfa80b4a20>
```

Note that `A(1)` has `(int,)` rather than `int` as its type parameter. It's not necessarily an issue, but it may be nice to simplify `(int,)` to `int`.

Reply:
I thought it would work for the `()` constructor too. But evidently I was wrong..

I addressed the issue.